### PR TITLE
[GEOS-9297] Add GetLegendGraphic request support on external WMS layers

### PR DESCRIPTION
**Pull request by imranrajjad**
-added GetLegendGraphic for Cascaded WMS Layer
-added integration tests
-updated docs

https://osgeo-org.atlassian.net/projects/GEOS/issues/GEOS-9297

### CODE REVIEW

**Comment by imranrajjad** [comment]
@nmco review please

**Comment by aaime** [comment]
@imranrajjad please keep the checklist intact :-)

**Comment by nmco** [comment]
On top of @aaime comment:
- you can ask for a reviewer using GitHub functionality with he same name :smile: 
- the PR name should be: `[GEOS-9297] Add GetLegendGraphic request support on external WMS layers`
- the commit name, you guess it, shoudl be: `[GEOS-9297] Add GetLegendGraphic request support on external WMS layers`
- when you start working on a jira ticket, always fill the fix version tag and mark it as in porgress

File: src/wms/src/main/java/org/geoserver/wms/CascadedLegendRequest.java [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+import net.sf.json.JSONArray;
+import net.sf.json.JSONObject;
+import org.geoserver.catalog.LegendInfo;
+import org.geoserver.catalog.impl.LegendInfoImpl;
+import org.geoserver.util.IOUtils;
+import org.geoserver.wms.GetLegendGraphicRequest.LegendRequest;
+import org.geoserver.wms.legendgraphic.JSONLegendGraphicBuilder;
+
+/** @author ImranR */
+// GetLegendGraphicRequest
```
[the remark this answers is not in the document]
Comment (imranrajjad) [reply src/wms/src/main/java/org/geoserver/wms/CascadedLegendRequest.java]: will remove it

File: doc/en/user/source/data/cascaded/wms.rst [hunk 2]
```diff
@@ -1,78 +1,76 @@
-.. _data_external_wms:
```
[the remark this answers is not in the document]
Comment (imranrajjad) [reply doc/en/user/source/data/cascaded/wms.rst]: > I just did an initial review with some general remarks, didn't cover all files. I tried this manually but it doesn't seems to work.
> I setup two GeoServer instance, 8060 and 8061, the second one contains the usual GeoServer release test data, and I configured a remote store for the second one using WMS 1.3.0 capabilities document. I published the remote layers on the first instance under the it.geosolutions workspace.
> Try to fetch legend for the default style using WMS 1.3.0, produced the following result:
> http://localhost:8060/geoserver/ows?service=WMS&version=1.3.0&request=GetLegendGraphic&format=image/png&width=20&height=20&layer=it.geosolutions:tiger_roads
> 
> <?xml version="1.0" encoding="UTF-8"?><ServiceExceptionReport version="1.3.0" xmlns="http://www.opengis.net/ogc" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.opengis.net/ogc http://localhost:8060/geoserver/schemas/wms/1.3.0/exceptions_1_3_0.xsd">   <ServiceException>
>       Can&apos;t obtain the schema for the required layer.
> Server does not specify a GetLegendGraphic operation. Cannot be performed
> </ServiceException></ServiceExceptionReport>
> ... using WMS 1.1.1:
> http://localhost:8060/geoserver/ows?service=wms&version=1.1.1&request=GetLegendGraphic&format=image/png&width=20&height=20&layer=it.geosolutions:tiger_roads
> 
> <?xml version="1.0" encoding="UTF-8" standalone="no"?><!DOCTYPE ServiceExceptionReport SYSTEM "http://localhost:8060/geoserver/schemas/wms/1.1.1/WMS_exception_1_1_1.dtd"> <ServiceExceptionReport version="1.1.1" >   <ServiceException>
>       Can&apos;t obtain the schema for the required layer.
> Server does not specify a GetLegendGraphic operation. Cannot be performed
> </ServiceException></ServiceExceptionReport>
> Can you give me some inputs on this, maybe I'm missing something  😨.

Can you please try this

http://cloudsdi.geo-solutions.it/geoserver/wms?service=wms&version=1.1.1&request=GetCapabilities

File: src/wms/src/main/java/org/geoserver/wms/CascadedLegendRequest.java [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+public class CascadedLegendRequest extends LegendRequest {
+
+    Logger LOGGER = Logger.getLogger(CascadedLegendRequest.class.getName());
+
+    GetLegendGraphicRequest request;
+
+    org.geotools.ows.wms.request.GetLegendGraphicRequest remoteLegendGraphicRequest;
+
+    JSONObject jsonLegend;
+
```
[the remark this answers is not in the document]
Comment (imranrajjad) [reply src/wms/src/main/java/org/geoserver/wms/CascadedLegendRequest.java]: understood

File: doc/en/user/source/data/cascaded/wms.rst (hunk 2 above)
[the remark this answers is not in the document]
Comment (aaime) [reply doc/en/user/source/data/cascaded/wms.rst]: Wondering if the client ends up using version 1.3.0 for some reason (e.g., negotiation happening regardless of the version number specified in the caps url)), since in 1.3.0 there is no indication of GetLegendGraphic in the wms 1.3.0 caps (formally because we do not support SLD 1.1 extensions, but only SLD 1.0 extensions, and GetLegendGraphic is one of them.... we should check if the sld 1.1/wms 1.3.0 version has anything substantially different)

**Comment by nmco** [comment]
The QA failure is valid, related with code formatting:
```
[INFO] Web Map Service Module ............................. FAILURE [02:02 min]
```

Can you please fix this @imranrajjad :smile: ?

File: doc/en/user/source/data/cascaded/wms.rst (hunk 2 above)
[the remark this answers is not in the document]
Comment (imranrajjad) [reply doc/en/user/source/data/cascaded/wms.rst]: This has been adressed by converting a getMap request of remote WMS service and turning it into a GetLegend Graphic. Now 1.3.0 version is also supported

**Comment by imranrajjad** [comment]
fixes in latest commit

- 1.3.0 GetLegendSupport / Tests
- Format Fixes

Can someone please re-run AppVeyor, seems like a false alert, the project failing test was not touched by this PR.
`[INFO] GeoWebCache (GWC) Module ........................... FAILURE [01:28 min]`

Thanks

File: src/wms/src/test/resources/org/geoserver/wms/caps130.xml [hunk 4]
```diff
@@ -1,86 +1,96 @@
...
-          <eastBoundLongitude>180</eastBoundLongitude>
-          <southBoundLatitude>-90</southBoundLatitude>
-          <northBoundLatitude>90</northBoundLatitude>
-        </EX_GeographicBoundingBox>
-        <BoundingBox CRS="EPSG:4326" minx="-90" miny="-180"
-          maxx="90" maxy="180" />
-      </Layer>
-    </Layer>
-  </Capability>
+	xmlns="http://www.opengis.net/wms"
```
Comment (nmco) [inline src/wms/src/test/resources/org/geoserver/wms/caps130.xml]: What's this change, unrelated format changes? If yes please remove them.

File: src/wms/src/main/java/org/geoserver/wms/legendgraphic/GetLegendGraphicKvpReader.java [hunk 5]
```diff
@@ -214,6 +227,50 @@ public GetLegendGraphicRequest read(Object req, Map kvp, Map rawKvp) throws Exce
...
+        WMSLayerInfo wmsLayerInfo = (WMSLayerInfo) layerInfo.getResource();
+        WMSStoreInfo wmsStoreInfo = wmsLayerInfo.getStore();
+        WebMapServer wmsServer = wmsStoreInfo.getWebMapServer(null);
+
+        org.geotools.ows.wms.request.GetLegendGraphicRequest remoteLegendGraphicRequest;
+
+        if (wmsServer.getCapabilities().getVersion().equalsIgnoreCase("1.3.0")) {
+            // WebMapServer under 1.3.0 version does not create GetLegendGraphicRequest
+            // since the XML tag <GetLegendGraphic> is not present under the <Capability>
+            // HACK HACK HACK
```
Comment (nmco) [inline src/wms/src/main/java/org/geoserver/wms/legendgraphic/GetLegendGraphicKvpReader.java]: ?

Comment (imranrajjad) [reply src/wms/src/main/java/org/geoserver/wms/legendgraphic/GetLegendGraphicKvpReader.java]: here we are checking if the WMS Server was created using 1.3.0 capabilities document, because that will fail `wmsServer.createGetLegendGraphicRequest()`. So we take a valid empty GetMap request and turn it into GetLegend request

File: src/wms/src/test/resources/org/geoserver/wms/caps130.xml (hunk 4 above)
Comment (imranrajjad) [reply src/wms/src/test/resources/org/geoserver/wms/caps130.xml]: looks like my eclipse is up to no good, I will fix this

**Comment by imranrajjad** [comment]
@nmco kindly review and merge

**Comment by nmco** [comment]
Looks good thx @imranrajjad.
